Feat/deployment creation

chrismclarke opened the pull request:
PR Checklist

- [x] - Latest `master` branch merged
- [x] - PR title descriptive (can be used in release notes)

## Breaking Changes
The active deployment json file has been renamed from `default.json` to `activeDeployment.json`. Any existing users will have an outstanding `default.json` file which should be deleted manually when checking out this PR or post merge.

## Description

Follow-up to #1756 and further refactor of work from #1755. This PR introduces a new workflow for creating deployments via the script `yarn workflow deployment create`. 

It also changes the way deployment configs are defined and extended, so that each deployment sits in its own folder even in cases of extended deployments. This is required so that extended deployments can still keep their own local copy of synced and processed sheets, which eventually would sync with their own git repos.

Main changes:
- Add `deployment create` workflow, including templates for new and extended deployments and gitignore file. 
- Improve the way deployment configs are tracked and processed. This should mean users are prompted less frequently to set their active deployment.
- Update deployment documentation
- Refactor all existing deployments to flat folder structures
- Tidy existing .gitignore files to allow deployments to specify their own rules.

## Git Issues

Closes #

## Review Notes
The pr impacts the deployment system, and so it would be good to double-check each of the existing deployments still works as expected (can set as active deployment, sync content and run as expected)

There is also scope if the changes made here are inconsistent with changes made in only deployment branches. 
I've tried to cross-check all deployment branches for cases where changes have been made to the config in the branch but not updated on master. The only changes I could find were efm, but would be good to have a double check. Additionally I also noted branches for aehub and fbm which don't have deployments (assume these branches can be deleted?)

The final check would be to review the added documentation and test the workflow for creating new and extending deployments

## Screenshots/Videos
**Updated deployment documentation**
![127 0 0 1_8000_developers_deployments_](https://user-images.githubusercontent.com/10515065/215641698-512a4bac-3609-4141-911d-7532cf5a17a7.png)

jfmcquade reviewed (approved):
Looks good. Putting aside the progress towards separate repos, I think this definitely makes working with deployments clearer, I like the use of "activeDeployment.json".

A few comments or points of clarification below:

### Creating a new deployment

The deployment folder for my newly created deployment ("my_deployment") is not tracked by git. Is this by design (we are working towards separate deployment repos after all)? If not, perhaps the folder path should be added as an exception to the `.idems_app/.gitignore` file as part of the deployment create workflow?
<img width="600" alt="Screenshot 2023-02-02 at 15 56 44" src="https://user-images.githubusercontent.com/64838927/216378418-1587d891-bef8-4b9f-9358-82e2bed8aa95.png">

### Documentation

The docs contain instructions on how to "Import Existing" deployments, but this feature is not yet implemented with this PR.

### Error handling

A couple of comments around error handling when setting and creating deployments. I don't think any of these are blocking issues, just things I noticed while testing.

- Previously if a deployment's config file didn't define a valid deployment, then the deployment would not be listed when running `yarn workflow deployment set`. Now the deployment name (folder name) can be selected, but doing so throws an error. This seems fine, although perhaps would be good to suppress the error that is logged below the "config not found" error.
  <img width="800" alt="Screenshot 2023-02-02 at 10 48 58" src="https://user-images.githubusercontent.com/64838927/216378645-ee438c65-43c8-4300-a1da-60c3f75f3c63.png">

- I can manually create a new deployment folder and provide an inadequate config file, and I am able to set this nominal deployment as active. An error is only thrown at the point of running `yarn start` to build the app. We could handle this error and alert the author that the issue is with the deployment config, although as authors are directed to create deployments using the workflow command, I don't think this is a problem.
  <img width="800" alt="Screenshot 2023-02-02 at 10 53 56" src="https://user-images.githubusercontent.com/64838927/216378780-72ea05f3-7158-4eda-bfdc-d9be4f0950d7.png">

esmeetewinkel reviewed (approved):
Functional test passed: I tested changing between existing deployments, they still work well, and creating new deployments either from existing or from scratch. 

When picking a deployment name with spaces `my new deployment` the name used was `my_new deployment`. I think this should either be `my_new_deployment`, or a warning should appear that deployment names cannot contain spaces.

<img width="615" alt="image" src="https://user-images.githubusercontent.com/74557272/216410865-993cd2d7-1a23-4599-a807-04ada9a8d30b.png">

chrismclarke commented on documentation/docs/developers/deployments.md:
```diff
@@ -1,63 +1,93 @@
 
 !!! tip
-    The folder can have any name, but it is recommended to avoid special characters and spaces to ensure compatibility across different devices and operating systems.
+    Extending deployments are more efficient if using content from the same google drive folder. Content can still be customised through the use of filters or other configuration options
```
As it happens I don't think it actually is (or at least not in the way it was originally written). 

When we kept extended deployments in the same folder as the original they shared the same cache of synced and converted google drive files, so it was very quick to reprocess when switching deployments. 

Now that every deployment has its own folder it also keeps its own cache of google drive folders, so no difference in efficiency. In the future we could consider a global google drive cache or making it easier to define multiple folders to sync just what is needed for a deployment. I'll remove the tip for now

chrismclarke commented:
Thanks for the inputs

@jfmcquade 
I've tidied the docs and missing config error, I also added a check for a specific `config.ts` file in each deployment folder before providing the option.
I couldn't think of a clever way to detect the config error you noted so I think for now reasonable to expect devs create deployment from the cli (or fix their own issues if breaking config)

@esmeetewinkel 
Thanks for finding the issue with the spaces, I've fixed to replace all spaces with underscores. I'm also going to add a follow-up PR to fix typos generally
